astroid/protocols.py

```diff
@@ -274,7 +274,14 @@ def _resolve_looppart(parts, assign_path, context):
 
 
 @decorators.raise_if_nothing_inferred
-def for_assigned_stmts(self, node=None, context=None, assign_path=None):
+def for_assigned_stmts(
+    self: Union[nodes.For, nodes.Comprehension],
+    node: Union[
+        nodes.List, nodes.Tuple, nodes.AssignName, nodes.AssignAttr, None
+    ] = None,
```

Comment:
How did you narrow it down to these types? I would need to spend more time looking at it.

Reply:
I looked at all calls to `assigned_stmts` and looked at what was provided as `node`. I'll show them here:

1)
https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/inference.py#L955
No `node`.

2)
https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/protocols.py#L294-L314
`node` is either `Tuple` or `List`.

3)
https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/protocols.py#L317-L322
`node` is either `AssignName` or `AssignAttr`

4)
https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/protocols.py#L420-L425
Just passes whatever `node` currently is -> no new types.

5)
https://github.com/PyCQA/astroid/blob/96ca4dc75aacaba4d1f4e9a66a72699c9fc93349/astroid/nodes/scoped_nodes.py#L803
No `node`.


I checked and `assigned_stmts` is not used in `pylint`.

Reply:
Beware that `assigned_stmts` is called on `self.parent` and not `self`. Thus I would think technically it can be `NodeNG`.

Reply:
> Beware that `assigned_stmts` is called on `self.parent` and not `self`. Thus I would think technically it can be `NodeNG`.

I have taken another look, but I'm not seeing how it can be anything other than the four types in there now. However, you have often proven me wrong so I won't claim I'm 100% sure. My thought process is as follows:
Case 1 and 5 can be disregarded immediately as they don't pass a `node`.
Case 4 does pass a `node`, but only passes the one already defined or gives the standard `None`. Because `assign_assigned_stmts` does nothing with `node` this doesn't allow any new types other than the ones possibly being passed to `assign_annassigned_stmts` to become `node`.
In case 2 and 3 only pass the `self` as `node`. Therefore we need only need to determine what their `self` can be to determine what `node` can be in any of the calls.
`sequence_assigned_stmts` (case 2) is never called directly and only a method on `Tuple` and `List`. Referring to your comment, we can only enter this function if `self.parent` is `Tuple` or `List`. I don't think it is possible to enter `sequence_assigned_stmts` with `self` not being `Tuple` or `List` and therefore the `node` created here is always `Tuple` or `List`.
A similar case can be made for `assend_assigned_stmts`. It is only accessed as method on `AssignName` and `AssignAttr`. `self` should always be either of them and therefore the `node` "created" in that function also needs to be.

Reply:
You are right! Thanks for the explanation. I missed that `self` is passed to node. That it's called on the parent doesn't change anything.